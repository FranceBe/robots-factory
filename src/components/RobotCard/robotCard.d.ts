import { ButtonProps } from 'components/Button/button'
import { IconProps } from 'components/Icon/icon'
import { LoadingBarProps } from 'components/LoadingBar/loadingBar'

export type RobotCardProps = {
  buttons: ButtonProps[]
  loadingBar: LoadingBarProps
  iconInfo?: IconProps
  textInfo: string
  robotId: string
}
