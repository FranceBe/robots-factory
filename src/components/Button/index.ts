import ButtonDefault, { Button } from 'components/Button/Button.component'

export { Button }

export default ButtonDefault
