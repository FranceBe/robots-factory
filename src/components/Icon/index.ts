import { FAIconByIconProps, Icon } from 'components/Icon/Icon.component'

export { FAIconByIconProps, Icon }
