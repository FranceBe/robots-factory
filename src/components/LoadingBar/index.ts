import { LoadingBar } from 'components/LoadingBar/LoadingBar.component'

export { LoadingBar }
