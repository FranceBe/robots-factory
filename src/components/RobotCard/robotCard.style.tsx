import styled from 'styled-components'
import { fontSizes, palette, spaces } from 'styles/variables'

export const CardBackground = styled.div`
  background-color: ${palette.white};
  padding: ${spaces.regular};
  border-radius: ${spaces.x_small};
  display: flex;
  align-items: center;
  width: fit-content;
`

export const RobotsContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  margin-right: ${spaces.regular};
  svg {
    margin-bottom: ${spaces.small};
  }
`

export const ButtonsAndLoadingContainer = styled.div``

export const ButtonsContainer = styled.div`
  display: flex;
  button {
    margin: ${spaces.small};
  }
`

export const LoadingAndInfoContainer = styled.div`
  margin: ${spaces.small};
`
export const InfoContainer = styled.div`
  min-height: ${fontSizes.medium};
  margin-top: ${spaces.small};
  display: flex;
  align-items: center;
  svg {
    margin-right: ${spaces.small};
  }
`
