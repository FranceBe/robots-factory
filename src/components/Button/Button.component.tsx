import { ButtonProps } from 'components/Button/button'
import { ButtonContainer, ContentContainer } from 'components/Button/button.style'
import { Icon } from 'components/Icon'
import React from 'react'
import ReactTooltip from 'react-tooltip'
import { palette } from 'styles/variables'
import { ThemeType } from 'utils/common.variables'

export const Button: React.FC<ButtonProps> = ({
  buttonType = ThemeType.primary,
  isActive = false,
  tooltipText,
  uniqueIndex = '1',
  ...rest
}) => {
  return (
    <div
      data-tip={tooltipText}
      // data-html and data-for are used
      // with ReactTooltip to allow html in tooltip and link
      // the div with the tooltip
      data-html={true}
      data-for={`button-tooltip${uniqueIndex}`}
      data-testid={'button-container'}
    >
      <ButtonContainer buttontype={ThemeType[buttonType]} {...rest}>
        <ContentContainer>{rest.children}</ContentContainer>
        {isActive && <Icon type={'robot'} />}
        {tooltipText && (
          <ReactTooltip
            html={true}
            backgroundColor={palette.white}
            textColor={palette.grey_primary}
            id={`button-tooltip${uniqueIndex}`}
          />
        )}
      </ButtonContainer>
    </div>
  )
}

export default Button
