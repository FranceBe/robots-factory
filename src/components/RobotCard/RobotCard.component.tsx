import { Button } from 'components/Button'
import { Icon } from 'components/Icon'
import { LoadingBar } from 'components/LoadingBar'
import {
  ButtonsAndLoadingContainer,
  ButtonsContainer,
  CardBackground,
  InfoContainer,
  LoadingAndInfoContainer,
  RobotsContainer,
} from 'components/RobotCard/robotCard.style'
import { RobotCardProps } from 'components/RobotCard/types/robotCard'
import React from 'react'

export const RobotCard: React.FC<RobotCardProps> = ({
  buttons,
  loadingBar,
  iconInfo,
  textInfo,
  robotId,
}) => {
  return (
    <CardBackground data-testid={'robot-card'}>
      <RobotsContainer>
        <Icon type={'robot'} size={'3x'} />
        <span>Robot {robotId}</span>
      </RobotsContainer>
      <ButtonsAndLoadingContainer>
        <ButtonsContainer>
          {buttons &&
            buttons.map((button, index) => <Button key={`button-${index}`} {...button} />)}
        </ButtonsContainer>
        <LoadingAndInfoContainer>
          <LoadingBar {...loadingBar} />
          <InfoContainer>
            {iconInfo ? <Icon {...iconInfo} /> : <Icon type={'spinner'} color={'transparent'} />}
            <span>{textInfo}</span>
          </InfoContainer>
        </LoadingAndInfoContainer>
      </ButtonsAndLoadingContainer>
    </CardBackground>
  )
}

export default RobotCard
