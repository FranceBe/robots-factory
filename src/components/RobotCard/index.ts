import { RobotCard } from 'components/RobotCard/RobotCard.component'

export { RobotCard }
