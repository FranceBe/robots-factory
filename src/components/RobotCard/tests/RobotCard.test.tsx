import '@testing-library/jest-dom/extend-expect'

import { render, screen } from '@testing-library/react'
import { ButtonProps } from 'components/Button/button'
import { IconProps } from 'components/Icon/icon'
import { LoadingBarProps } from 'components/LoadingBar/loadingBar'
import { RobotCard } from 'components/RobotCard'
import React from 'react'
import { ThemeType } from 'utils/common.variables'

describe('RobotCard component', () => {
  // Set up test
  const buttons: ButtonProps[] = [
    { buttonType: ThemeType.primary, children: 'Miner Foo', isActive: true },
    { buttonType: ThemeType.primary, children: 'Miner Bar' },
    {
      buttonType: ThemeType.primary,
      children: 'Assembler Foobar',
      disabled: true,
    },
    {
      buttonType: ThemeType.secondary,
      children: 'Acheter un robot',
      disabled: true,
    },
  ]
  const loadingBar: LoadingBarProps = { timeBase: 1, timeLeft: 0.5 }
  const iconInfo: IconProps = { spin: true, type: 'spinner' }
  const textInfo = 'Le robot est en train de miner du Foo. Cela prend 1s / Foo.'
  const robotId = '1'
  const initTest = () =>
    render(
      <RobotCard
        buttons={buttons}
        loadingBar={loadingBar}
        iconInfo={iconInfo}
        textInfo={textInfo}
        robotId={robotId}
      />,
    )

  it('should match snapshot', () => {
    const { container } = initTest()

    expect(container.firstChild).toMatchSnapshot()
  })
  it('should render 3 icons and robot name', () => {
    initTest()

    const icons = screen.getAllByRole('img', { hidden: true })
    // 2 robot icons (1 with name and 1 in the active button)
    // 1 spinner icon
    expect(icons).toHaveLength(3)
    expect(icons[0]).toHaveClass('fa-robot')
    expect(icons[1]).toHaveClass('fa-robot')
    expect(icons[2]).toHaveClass('fa-spinner')

    const robotName = `Robot ${robotId}`
    expect(screen.getByText(robotName)).toBeInTheDocument()
  })
  it('should render buttons.length button', () => {
    initTest()

    const buttonsRendered = screen.getAllByRole('button')
    expect(buttonsRendered).toHaveLength(buttons.length)
    buttons.forEach((button) => expect(screen.getByText(`${button.children}`)).toBeInTheDocument())
  })
  it('shoud render a LoadingBar', () => {
    initTest()

    expect(screen.getByTestId('loading-bar')).toBeInTheDocument()
    expect(screen.getByTestId('filled-bar')).toBeInTheDocument()
  })
  it('should render textInfo', () => {
    initTest()

    expect(screen.getByText(textInfo)).toBeInTheDocument()
  })
})
