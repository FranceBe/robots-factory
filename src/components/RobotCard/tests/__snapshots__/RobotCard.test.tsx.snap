// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RobotCard component should match snapshot 1`] = `
<div
  class="sc-gKAaRy hQwNST"
  data-testid="robot-card"
>
  <div
    class="sc-iCoGMd hDMOLj"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-robot fa-w-20 fa-3x "
      data-icon="robot"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 640 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M32,224H64V416H32A31.96166,31.96166,0,0,1,0,384V256A31.96166,31.96166,0,0,1,32,224Zm512-48V448a64.06328,64.06328,0,0,1-64,64H160a64.06328,64.06328,0,0,1-64-64V176a79.974,79.974,0,0,1,80-80H288V32a32,32,0,0,1,64,0V96H464A79.974,79.974,0,0,1,544,176ZM264,256a40,40,0,1,0-40,40A39.997,39.997,0,0,0,264,256Zm-8,128H192v32h64Zm96,0H288v32h64ZM456,256a40,40,0,1,0-40,40A39.997,39.997,0,0,0,456,256Zm-8,128H384v32h64ZM640,256V384a31.96166,31.96166,0,0,1-32,32H576V224h32A31.96166,31.96166,0,0,1,640,256Z"
        fill="currentColor"
      />
    </svg>
    <span>
      Robot 
      1
    </span>
  </div>
  <div
    class="sc-fujyAs"
  >
    <div
      class="sc-pNWdM ecOhoC"
    >
      <div
        data-for="button-tooltip1"
        data-html="true"
        data-testid="button-container"
      >
        <button
          buttontype="primary"
          class="sc-bdnxRM kHkzzW"
        >
          <div
            class="sc-gtsrHT FiSsO"
          >
            Miner Foo
          </div>
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-robot fa-w-20 "
            data-icon="robot"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 640 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M32,224H64V416H32A31.96166,31.96166,0,0,1,0,384V256A31.96166,31.96166,0,0,1,32,224Zm512-48V448a64.06328,64.06328,0,0,1-64,64H160a64.06328,64.06328,0,0,1-64-64V176a79.974,79.974,0,0,1,80-80H288V32a32,32,0,0,1,64,0V96H464A79.974,79.974,0,0,1,544,176ZM264,256a40,40,0,1,0-40,40A39.997,39.997,0,0,0,264,256Zm-8,128H192v32h64Zm96,0H288v32h64ZM456,256a40,40,0,1,0-40,40A39.997,39.997,0,0,0,456,256Zm-8,128H384v32h64ZM640,256V384a31.96166,31.96166,0,0,1-32,32H576V224h32A31.96166,31.96166,0,0,1,640,256Z"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
      <div
        data-for="button-tooltip1"
        data-html="true"
        data-testid="button-container"
      >
        <button
          buttontype="primary"
          class="sc-bdnxRM kHkzzW"
        >
          <div
            class="sc-gtsrHT FiSsO"
          >
            Miner Bar
          </div>
        </button>
      </div>
      <div
        data-for="button-tooltip1"
        data-html="true"
        data-testid="button-container"
      >
        <button
          buttontype="primary"
          class="sc-bdnxRM kHkzzW"
          disabled=""
        >
          <div
            class="sc-gtsrHT FiSsO"
          >
            Assembler Foobar
          </div>
        </button>
      </div>
      <div
        data-for="button-tooltip1"
        data-html="true"
        data-testid="button-container"
      >
        <button
          buttontype="secondary"
          class="sc-bdnxRM kdGdbW"
          disabled=""
        >
          <div
            class="sc-gtsrHT FiSsO"
          >
            Acheter un robot
          </div>
        </button>
      </div>
    </div>
    <div
      class="sc-jrsJWt hKtNJe"
    >
      <div
        class="sc-dlnjwi gQEkhQ"
        data-testid="loading-bar"
      >
        <div
          class="sc-hKFxyN bmKIvQ"
        >
          <div
            class="sc-jSFjdj gxmQZS"
            data-testid="filled-bar"
            filled="50"
          />
        </div>
        <div
          class="sc-eCApnc bChZKY"
        >
          0.5
           s
        </div>
      </div>
      <div
        class="sc-kEqXSa kLYlHg"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-spinner fa-w-16 fa-spin "
          data-icon="spinner"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M304 48c0 26.51-21.49 48-48 48s-48-21.49-48-48 21.49-48 48-48 48 21.49 48 48zm-48 368c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zm208-208c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.49-48-48-48zM96 256c0-26.51-21.49-48-48-48S0 229.49 0 256s21.49 48 48 48 48-21.49 48-48zm12.922 99.078c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.491-48-48-48zm294.156 0c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48c0-26.509-21.49-48-48-48zM108.922 60.922c-26.51 0-48 21.49-48 48s21.49 48 48 48 48-21.49 48-48-21.491-48-48-48z"
            fill="currentColor"
          />
        </svg>
        <span>
          Le robot est en train de miner du Foo. Cela prend 1s / Foo.
        </span>
      </div>
    </div>
  </div>
</div>
`;
